### ISSUE TITLE
Memory Leak using dbt rpc in 0.19.0

### ISSUE THREAD

**Issue by robertdefilippi** [fishtown-analytics/dbt#3412]
### Describe the bug
Hi,

We're running a dbt rpc server, and after a few days our server will fail due to memory issues. Because we're only running dbt through the Fishtown docker image, I'm not sure there would be another culprit for the memory leak. I'm not sure if this is a known bug already or if someone else has reported it, but I wanted to bring it up. I'm also not sure I provided enough information, so if there are any additional pieces of information I can provide please let me know.

### Expected behavior
My understanding is we should be able to have the rpc running for the life of the server without having a memory leak.

### Screenshots and log output

Here is an image of our dbt rpc server over time. Each of the lines represents how much memory the server is using. As you can see after each of the daily runs, the total memory on the server keeps increasing.

![image](https://user-images.githubusercontent.com/9288605/120335243-54ab8100-c2bf-11eb-8c39-ff474f1f2cf5.png)

### System information
**Which database are you using dbt with?**
- [ ] postgres
- [x] redshift
- [ ] bigquery
- [ ] snowflake
- [ ] other (specify: ____________)

**The output of `dbt --version`:**
We're using this version of dbt:
https://hub.docker.com/layers/fishtownanalytics/dbt/0.19.0/images/sha256-99b5bfb6cee7a48a27b2ac6531622ceedbbcc3f27742f879344a7698dcb51bae?context=explore

**The operating system you're using:**
Above ^^

**The output of `python --version`:**
Above ^^

### Additional context
If there are any additional pieces of information or something I missed that will help the debugging process, please let me know.

**Comment by jtcohen6** [comment]
@robertdefilippi Thanks for opening the issue! I can't know exactly what's causing the memory leakage, but I have a guess: If you're submitting a high volume of tasks, the logs and results for those tasks are persisted in memory. For this reason, the RPC server has a GarbageCollector, which removes finished tasks from memory. Its default behavior is to clear out logs for tasks once they are too old (>30 days) or too numerous (>1000):

https://github.com/fishtown-analytics/dbt/blob/1e5a7878e5188fa58f91e2f257041f28fdb948fc/core/dbt/rpc/gc.py#L27-L32

It's possible that you're hitting memory limits before hitting either of the triggers for automated garbage collection. You could use the [`ps`](https://docs.getdbt.com/reference/commands/rpc#ps) method to list all running and completed processes; my guess is that the number of processes should scale linearly with the memory usage you're seeing above.

There isn't any way today to change the default limits used by automated garbage collection. There is, however, an undocumented `gc` method you can use, which accepts parameters for `task_ids`, `before`, and `settings`:

https://github.com/fishtown-analytics/dbt/blob/1e5a7878e5188fa58f91e2f257041f28fdb948fc/core/dbt/rpc/builtins.py#L48-L61

Given their types, it's not... totally... clear to me how you could pass `before` (`Optional[datetime]`) and `settings` (`Optional[GCSettings]`) via JSON request. It doesn't look like we have logic to deserialize them out of corresponding JSON objects.

Your best bet today may be to encode some custom logic to find old tasks (via `ps`) and pass as a list of `task_ids` to the `gc` method. I'd also be open to a contribution that improves user-side control over the `gc` method params, or (better yet) the default garbage collection settings.

**Comment by robertdefilippi** [comment]
Hey @jtcohen6 this is some great digging! Thanks for showing me these methods and I think we should be able to solve the issue on our end with this information.

If we _do_ end up doing that, I should probably update the documentation for [this method](https://github.com/fishtown-analytics/dbt/blob/1e5a7878e5188fa58f91e2f257041f28fdb948fc/core/dbt/rpc/gc.py#L112). Are there any additional contribution guidelines for documenting a methods specific to the rpc section of dbt?

Thanks :)

**Comment by jtcohen6** [comment]
> Are there any additional contribution guidelines for documenting a methods specific to the rpc section of dbt?

Not really! Today, all documentation for RPC methods is here: https://github.com/fishtown-analytics/docs.getdbt.com/blob/current/website/docs/reference/commands/rpc.md

Given that the `gc` method isn't documented _at all_ today, if you manage to use it with some success, I would welcome a PR there to add it in
